src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionFactory.cs

```diff
@@ -217,16 +217,26 @@ internal bool TryGetConnection(DbConnection owningConnection, TaskCompletionSour
 
                             // now that we have an antecedent task, schedule our work when it is completed.
                             // If it is a new slot or a completed task, this continuation will start right away.
-                            newTask = s_pendingOpenNonPooled[idx].ContinueWith((_) =>
+                            newTask = s_pendingOpenNonPooled[idx].ContinueWith((_, state) =>
                             {
-                                var newConnection = CreateNonPooledConnection(owningConnection, poolGroup, userOptions);
-                                if ((oldConnection != null) && (oldConnection.State == ConnectionState.Open))
+                                // Avoid implicitly closing over our variables to avoid a closure allocation
+                                var tuple = (Tuple<DbConnectionFactory, DbConnection, DbConnectionPoolGroup, DbConnectionOptions, DbConnectionInternal>)state;
+                                DbConnectionFactory self = tuple.Item1;
+                                DbConnection capturedOwningConnection = tuple.Item2;
+                                DbConnectionPoolGroup capturedPoolGroup = tuple.Item3;
+                                DbConnectionOptions capturedUserOptions = tuple.Item4;
+                                DbConnectionInternal capturedOldConnection = tuple.Item5;
+                                
+                                var newConnection = self.CreateNonPooledConnection(capturedOwningConnection, capturedPoolGroup, capturedUserOptions);
+                                if ((capturedOldConnection != null) && (capturedOldConnection.State == ConnectionState.Open))
                                 {
-                                    oldConnection.PrepareForReplaceConnection();
-                                    oldConnection.Dispose();
+                                    capturedOldConnection.PrepareForReplaceConnection();
+                                    capturedOldConnection.Dispose();
                                 }
                                 return newConnection;
-                            }, cancellationTokenSource.Token, TaskContinuationOptions.LongRunning, TaskScheduler.Default);
+                            },
+                            Tuple.Create(this, owningConnection, poolGroup, userOptions, oldConnection),
```

Comment:
The state passed to a task is exposed from its AsyncState property.  If this task is returned out to external code, that code will have access to all of this state.  Do you know that this is safe?  Same question for all cases of this.